### PR TITLE
Generic Record Decoder improvements, should take into account the schema field type

### PR DESCRIPTION
**Pull request by pomadchin**
This PR is a followup for #701 and covers a wonderful suggestion by @googley42 (https://github.com/zio/zio-dynamodb/pull/701#discussion_r2324303331)

All tests pass and also tried internally! We probabable need to add tests for all of the edge cases though :D

### CODE REVIEW

File: dynamodb/src/main/scala/zio/dynamodb/Codec.scala [hunk 1]
```diff
@@ -575,15 +575,25 @@ private[dynamodb] object Codec {
...
+                  ContainerField.containerField(schema) match {
+                    case ContainerField.Optional => Right(k -> None)
+                    case ContainerField.Chunk    => Right(k -> Chunk.empty)
+                    case ContainerField.Sequence => Right(k -> List.empty)
+                    case ContainerField.Map      => Right(k -> Map.empty)
+                    case ContainerField.Set      => Right(k -> Set.empty)
+                    case ContainerField.Scalar   => errorOrValue
+                  }
+                else
+                  errorOrValue
```
Comment (pomadchin) [inline dynamodb/src/main/scala/zio/dynamodb/Codec.scala]: This will 100% make the library user experience better; thx @googley42!

Comment (googley42) [reply dynamodb/src/main/scala/zio/dynamodb/Codec.scala]: actually my comment was not substantiated by any unit tests. I think generic record processing may be different from case class processing so we could be OK - just need some tests to prove it

File: dynamodb/src/test/scala/zio/dynamodb/codec/CodecRoundTripSpec.scala [hunk 2]
```diff
@@ -395,6 +395,41 @@ object CodecRoundTripSpec extends ZIOSpecDefault with CodecTestFixtures {
...
+      val case21List = Case21List(
+        "id", "f0", "f1", "f2", "f3", "f4", "f5", "f6", "f7", "f8", "f9", "f10",
+        "f11", "f12", "f13", "f14", "f15", "f16", "f17", Nil, Nil
+      )
+      // format: on
+
+      val itemCase21Option  = DynamoDBQuery.toItem(case21Option)
+      val itemCase21Decoded = DynamoDBQuery.fromItem[Case21List](itemCase21Option)
+      assert(itemCase21Decoded)(isRight(equalTo(case21List)))
+    },
```
Comment (pomadchin) [inline dynamodb/src/test/scala/zio/dynamodb/codec/CodecRoundTripSpec.scala:415]: I wonder if that's a not desired behavior; Nil's are encoded as the actual empty fields in the Dynamo item; technically a mistake.

File: dynamodb/src/test/scala/zio/dynamodb/codec/CodecRoundTripSpec.scala [hunk 3]
```diff
@@ -395,6 +396,36 @@ object CodecRoundTripSpec extends ZIOSpecDefault with CodecTestFixtures {
...
+        "id", "f0", "f1", "f2", "f3", "f4", "f5", "f6", "f7", "f8", "f9", "f10",
+        "f11", "f12", "f13", "f14", "f15", "f16", "f17", None, None,
+        None, None, None, None, None, None, None, None, None, None, None
+      )
+      // format: on
+
+      val itemBigOption  = DynamoDBQuery.toItem(bigOption)
+      val bigListDecoded = DynamoDBQuery.fromItem[BigList](itemBigOption)
+
+      assert(bigListDecoded)(isLeft(equalTo(DecodingError(message = "field 'f18' not found in AttributeValue map"))))
```
Comment (pomadchin) [inline dynamodb/src/test/scala/zio/dynamodb/codec/CodecRoundTripSpec.scala:428]: encoding option => no field in the doc; on decoding the list is non optional => the field not found error makes sense; @googley42 ^ case classes suffer the same feature.

**Comment by googley42** [comment]
apologies for delay!

One gap: for `Chunk` / `Sequence` / `Map` / `Set`, this currently falls through to `errorOrValue`, so a missing field decodes as an error rather than an empty collection. `decodeFields`  
(the case-class path, [`Codec.scala#L1092-L1098`](https://github.com/zio/zio-dynamodb/blob/series/2.x/dynamodb/src/main/scala/zio/dynamodb/Codec.scala#L1092-L1098)) already handles all   
five `ContainerField` cases consistently — added by #701 — so I think `genericRecordDecoder` should mirror that match in full rather than special-casing just `Optional`:

```scala                                                                                                                                                                                   
case None =>                                                                                                                                                                               
  ContainerField.containerField(schema) match {                                                                                                                                            
    case ContainerField.Optional => Right(key -> None)                                                                                                                                     
    case ContainerField.Chunk    => Right(key -> Chunk.empty)                                                                                                                              
    case ContainerField.Sequence => Right(key -> List.empty)                                                                                                                               
    case ContainerField.Map      => Right(key -> Map.empty)                                                                                                                                
    case ContainerField.Set      => Right(key -> Set.empty)                                                                                                                                
    case ContainerField.Scalar   => Left(DecodingError(s"field '$key' not found in AttributeValue map"))                                                                                   
  }                                                                                                                                                                                        
```                                                                                                                                                                                        

No new machinery needed — this reuses `ContainerField`/`containerField(schema)` exactly as `decodeFields` already does.

Worth flagging: this isn't just a dynamic-schema edge case. `zio-schema`'s own derivation macro falls back to `Schema.GenericRecord` for any `case class` past 22 fields (`CaseClassN`     
stops at 22), so `Big`/`BigList` in this PR's own test are already exercising exactly that path — any ordinary wide case class hits `genericRecordDecoder` transparently, no               
`Schema.record` required.

Which also means the `BigOption` → `BigList` test in this PR needs its assertion flipped, not just left as-is: it currently expects

```scala                                                                                                                                                                                   
isLeft(equalTo(DecodingError(message = "field 'f18' not found in AttributeValue map")))                                                                                                    
```                                                                                                                                                                                        

for a missing `Sequence`-typed field, but under the full fix that becomes `isRight(...)` with `Nil` in `f18`/`f19` — matching what `Case21Option` → `Case21List` (the `Optional` case, just
above it) already asserts. Also still missing: a `Chunk` and a `Map`/`Set` case, neither of which any existing test (this PR, #701, or the property-based `SchemaGen`/`CodecRoundTripSpec`
generators) currently covers for a missing field on a `GenericRecord`-routed type.

BTW I have a PR here https://github.com/zio/zio-dynamodb/pull/915 that implements all the features discussed

**Comment by pomadchin** [comment]
Yes it makes sense; I'll close my PR than for now. Thanks!
